addons/website_links/static/src/interactions/website_links_charts.js

```diff
@@ -1,287 +1,226 @@
-import { loadBundle } from "@web/core/assets";
-import { deserializeDate } from "@web/core/l10n/dates";
+import { Interaction } from "@web/public/interaction";
 import { _t } from "@web/core/l10n/translation";
-import publicWidget from "@web/legacy/js/public/public_widget";
+import { deserializeDate } from "@web/core/l10n/dates";
+import { registry } from "@web/core/registry";
+import { loadBundle } from "@web/core/assets";
 const { DateTime } = luxon;
 
-var BarChart = publicWidget.Widget.extend({
-    /**
-     * @constructor
-     * @param {Object} parent
-     * @param {Object} beginDate
-     * @param {Object} endDate
-     * @param {Object} dates
-     */
-    init: function (parent, beginDate, endDate, dates) {
-        this._super.apply(this, arguments);
-        this.beginDate = beginDate.startOf("day");
-        this.endDate = endDate.startOf("day");
-        if (this.beginDate.toISO() === this.endDate.toISO()) {
-            this.endDate = this.endDate.plus({ days: 1 });
+class WebsiteLinksCharts extends Interaction {
+    static selector = ".o_website_links_chart";
+    dynamicContent = {};
+
+    setup() {
+        this.orm = this.services.orm;
+        const linkId = parseInt(this.el.querySelector("#link_id").value);
+        this.linksDomain = ["link_id", "=", linkId];
+    }
+
+    async willStart() {
+        const totalClicks = await this.totalClicks();
+        const clicksByDay = await this.clicksByDay();
+        const clicksByCountry = await this.clicksByCountry();
+        const lastWeekClicksByCountry = await this.lastWeekClicksByCountry();
+        const lastMonthClicksByCountry = await this.lastMonthClicksByCountry();
+
+        if (!totalClicks) {
+            const message = _t("There is no data to show");
+            this.el.querySelector("#all_time_charts").prepend(message);
+            this.el.querySelector("#last_month_charts").prepend(message);
+            this.el.querySelector("#last_week_charts").prepend(message);
+            return;
```

Comment:
We cannot modify the DOM inside `willStart`. This would cause problems in the future e.g. if that interaction runs in website edit mode: it would impact the page history, while the changes are not done by the user.

(Just thinking out loud here) Not sure why `prepend` was used in this case, I guess it cannot be replaced by a `dynamicContent` entry, that would destroy the divs inside.

So, just store the values in `this` in `willStart` and update the DOM in `start`.
And in that case, we've actually tried to define all variables in the `setup` method so that it can be read as a kind of documentation.